src/thunderbots/software/multithreading/thread_safe_buffer.h

```diff
@@ -0,0 +1,157 @@
+
+    explicit ThreadSafeBuffer(std::size_t buffer_size);
+
+    // Copying this class is not permitted
+    ThreadSafeBuffer(const ThreadSafeBuffer&) = delete;
+
+    /**
+     * Removes the least recently value added to the buffer and returns it
+     *
+     * If the buffer is empty, this function will *block* until:
+     * - a value becomes available
+     * - the given amount of time is exceeded
+     * - the destructor of this class is called
+     *
+     * @param max_wait_time The maximum duration to wait for a new value before
+     *                      returning
+     *
+     * @return The most recently value added to the buffer, or std::nullopt if none is
+     *         available
+     */
+    std::optional<T> pullLeastRecentlyAddedValue(
+        Duration max_wait_time = Duration::fromSeconds(0));
+
+    /**
+     * Removes the most recently value added to the buffer and returns it
+     *
+     * If the buffer is empty, this function will *block* until:
+     * - a value becomes available
+     * - the given amount of time is exceeded
+     * - the destructor of this class is called
+     *
+     * @param max_wait_time The maximum duration to wait for a new value before
+     *                      returning
+     *
+     * @return The most recently value added to the buffer, or std::nullopt if none is
+     *         available
+     */
+    std::optional<T> pullMostRecentlyAddedValue(
+        Duration max_wait_time = Duration::fromSeconds(0));
+
+    /**
+     * Push the given value onto the buffer
+     *
+     * If the buffer is already full, this will overwrite the least recently added value
+     *
+     * @param value The value to push onto the buffer
+     */
+    void push(const T& value);
+
+    ~ThreadSafeBuffer();
+
+   private:
+    /**
+     * Waits for the buffer to have at least one value
+     *
+     * @param max_wait_time The maximum duration to wait for a new value before
+     *                      returning
+     *
+     * @return A unique_lock on the contents of the buffer
+     */
+    std::unique_lock<std::mutex> waitForBufferToHaveAValue(Duration max_wait_time);
+
+    std::mutex buffer_mutex;
+    boost::circular_buffer<T> buffer;
+
+    std::condition_variable received_new_value;
+
+    bool destructor_called;
+};
+
+template <typename T>
+ThreadSafeBuffer<T>::ThreadSafeBuffer(std::size_t buffer_size)
+    : buffer(buffer_size), destructor_called(false)
+{
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::pullLeastRecentlyAddedValue(Duration max_wait_time)
+{
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.front();
+        buffer.pop_front();
+    }
+    return result;
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::pullMostRecentlyAddedValue(Duration max_wait_time)
+{
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
```

Comment:
Looks like a scoped lock to me, would like confirmation of this understanding

Reply:
so it's a unique lock, but looks like when it loses scope destructor will be implicitly called and that will unlock the mutex

Reply:
Yup, basically all correct. Added the following comment in both cases where we do this to clarify:
```
    // We hold the returned lock in a variable here so that we hold the lock on the
    // buffer mutex until the lock is destructed at the end of this function
```